### PR TITLE
[FEATURES] Clean up Data Operation after finished

### LINKED ISSUE #3415: [FEATURES] Clean up Data Operation after finished
**What feature you'd like to add:**

Delete Data Operation(Dataload/DataProcess/DataMigrate) from Kubernetes by using `ttlSecondsAfterFinished ` after it has finished for a (configurable) period of time. 

**Why is this feature needed:**

### PR DESCRIPTION
**Pull request by wangshli**
### Ⅰ. Describe what this PR does
Clean up Data Operation after `ttlSecondsAfterFinished` seconds after data operation completion.

### Ⅱ. Does this pull request fix one issue?

fixes #3415 

### Ⅲ. List the added test cases (unit test/integration test) if any, please explain if no tests are needed.
test/prow/testcases/common/data_operation_clean_up.py

### Ⅴ. Special notes for reviews

### CODE REVIEW

File: pkg/ddc/base/operation.go [hunk 1]
```diff
@@ -238,6 +238,19 @@ func (t *TemplateEngine) reconcileComplete(ctx cruntime.ReconcileRequestContext,
 		ctx.Recorder.Eventf(object, v1.EventTypeNormal, common.DataOperationSucceed,
 			"%s %s succeeded", operation.GetOperationType(), object.GetName())
 	}
+
+	// 5. clean up data operation
+	if len(opStatusToUpdate.Conditions) > 0 {
```
Comment (cheyang) [inline pkg/ddc/base/operation.go:243]: Can we use more reliable logic to handle the TTL situation? And should this logic be put in the end of the fuction?

Comment (wangshli) [reply pkg/ddc/base/operation.go]: put the clean up in the end of reconcileComplete due to the following two considerations:
1. If clean up in the beginning of `reconcileComplete`, in the first `reconcileComplete` the job completion time has not  updated, so the clean up actually do in the following `reconcileComplete`, and this need to add requeue in the end. But if the data operation do not set the `ttlSecondsAfterFinished `, the reque is useless.
2. If data operation set `ttlSecondsAfterFinished`, the `CleanUpIfArriveTTL` would calculate the remaining time to clean up, and requeue after this remaining time, and there will be only next one `reconcileComplete` to clean up the data operation. The `reconcileComplete` only execute two times and there is only small overhead.

File: pkg/ddc/base/clean_up.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+import (
+	"fmt"
+	"time"
+
+	datav1alpha1 "github.com/fluid-cloudnative/fluid/api/v1alpha1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// GetRemainingTimeToCleanUp return not nil remaining time if data operation has completion time and set ttlAfterFinished
+func GetRemainingTimeToCleanUp(object client.Object, opStatus *datav1alpha1.OperationStatus, operateType datav1alpha1.OperationType) (*time.Duration, error) {
```
Comment (cheyang) [inline pkg/ddc/base/clean_up.go:12]: How about naming the function to Timeleft,  then put it into utils package.

File: pkg/ddc/base/clean_up.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	if err != nil {
+		return nil, err
+	}
+	if ttl == nil {
+		// if data operation does not set ttlAfterFinished, return nil
+		return nil, nil
+	}
+
+	curTime := time.Now()
+	completionTime := opStatus.Conditions[0].LastProbeTime
```
Comment (cheyang) [inline pkg/ddc/base/clean_up.go:28]: Why the completionTIme is lastProbeTime?

Comment (wangshli) [reply pkg/ddc/base/clean_up.go:28]: Currently, we have no `jobCompletionTime` so I use `lastProbeTime` instead. And when job complete, it is actually same as `jobCompletionTime`.

Comment (cheyang) [reply pkg/ddc/base/clean_up.go:28]: Can we avoid using conditions[0] which is not sure  for jobCompletionTime?

Comment (wangshli) [reply pkg/ddc/base/clean_up.go:28]: Now, data operation will be cleaned up regardless of complete or failure if ttlAfterFinished has been set. Do you mean we only clean up completed data operation?

Comment (cheyang) [reply pkg/ddc/base/clean_up.go:28]: I mean only clean up completed or failed data operation.

Comment (wangshli) [reply pkg/ddc/base/clean_up.go:28]: ok.

File: pkg/ddc/base/clean_up.go (hunk 2 above)
Comment (wangshli) [reply pkg/ddc/base/clean_up.go]: done.

File: pkg/utils/dataoperation.go [hunk 4]
```diff
@@ -113,3 +115,80 @@ func HasPrecedingOperation(obj client.Object) (has bool, err error) {
...
+		if !ok {
+			return ttl, fmt.Errorf("object %v is not a DataBackup", object)
+		}
+		ttl = dataBackup.Spec.TTLSecondsAfterFinished
+	case datav1alpha1.DataLoadType:
+		dataLoad, ok := object.(*datav1alpha1.DataLoad)
+		if !ok {
+			return ttl, fmt.Errorf("object %v is not a DataLoad", object)
+		}
+		if dataLoad.Spec.Policy == datav1alpha1.Cron {
```
Comment (xliuqq) [inline pkg/utils/dataoperation.go:172]: how about let this ttl-get logic be a method of `dataoperation.OperationInterface`, remove these `case` judgement ?

File: test/prow/testcases/common/data_operation_clean_up.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+    dataset = fluidapi.assemble_dataset("alluxio-webufs") \
+        .set_namespaced_name(namespace, name) \
+        
+    runtime = fluidapi.assemble_runtime("alluxio-webufs") \
+        .set_namespaced_name(namespace, name) \
+        .set_tieredstore(mediumtype="MEM", path="/dev/shm", quota="2Gi")
+    
+    dataload_name = "test-dataload"
+    cron_dataload = fluidapi.DataLoad(name=dataload_name, namespace=namespace) \
```
Comment (xliuqq) [inline test/prow/testcases/common/data_operation_clean_up.py:104]: it is not cron dataload, rename it.

File: pkg/ddc/base/operation.go [hunk 6]
```diff
@@ -238,13 +259,40 @@ func (t *TemplateEngine) reconcileComplete(ctx cruntime.ReconcileRequestContext,
 		ctx.Recorder.Eventf(object, v1.EventTypeNormal, common.DataOperationSucceed,
 			"%s %s succeeded", operation.GetOperationType(), object.GetName())
 	}
```
Comment (TrafalgarZZZ) [inline pkg/ddc/base/operation.go:261]: Better to remove the "no need to requeue" log message in Line 258

File: pkg/ddc/base/operation.go [hunk 7]
```diff
@@ -190,6 +191,26 @@ func (t *TemplateEngine) reconcileComplete(ctx cruntime.ReconcileRequestContext,
...
+	var remaining *time.Duration
+	if utils.NeedCleanUp(object, opStatus, operation.GetOperationType()) {
+		remaining, err := utils.Timeleft(object, opStatus, operation.GetOperationType())
+		if err != nil {
+			log.Error(err, fmt.Sprintf("Failed to get remaining time to clean up for operation %s", operation.GetOperationType()))
+			return utils.RequeueIfError(err)
+		}
+		// if remaining time to clean up is not nil and less than 0, clean up data operation
+		if remaining != nil && *remaining <= 0 {
+			if err = ctx.Client.Delete(context.TODO(), object); err != nil {
```
Comment (TrafalgarZZZ) [inline pkg/ddc/base/operation.go:204]: Please ignore NotFound error here.

File: pkg/ddc/base/operation.go [hunk 8]
```diff
@@ -277,5 +325,11 @@ func (t *TemplateEngine) reconcileFailed(ctx cruntime.ReconcileRequestContext, o
 		log.Info(fmt.Sprintf("%s failed, won't requeue", operation.GetOperationType()))
 		ctx.Recorder.Eventf(object, v1.EventTypeWarning, common.DataOperationFailed, "%s %s failed", operation.GetOperationType(), object.GetName())
 	}
+
```
Comment (TrafalgarZZZ) [inline pkg/ddc/base/operation.go:328]: Better to remove the "won't requeue" log message in Line 325

File: pkg/ddc/base/operation.go [hunk 9]
```diff
@@ -238,13 +259,40 @@ func (t *TemplateEngine) reconcileComplete(ctx cruntime.ReconcileRequestContext,
...
+	var remaining *time.Duration
+	if utils.NeedCleanUp(object, opStatus, operation.GetOperationType()) {
+		remaining, err := utils.Timeleft(object, opStatus, operation.GetOperationType())
+		if err != nil {
+			log.Error(err, fmt.Sprintf("Failed to get remaining time to clean up for operation %s", operation.GetOperationType()))
+			return utils.RequeueIfError(err)
+		}
+		// if remaining time to clean up is not nil and less than 0, clean up data operation
+		if remaining != nil && *remaining <= 0 {
+			if err = ctx.Client.Delete(context.TODO(), object); err != nil {
```
Comment (TrafalgarZZZ) [inline pkg/ddc/base/operation.go:286]: Ditto. Ignore NotFound error

Comment (wangshli) [reply pkg/ddc/base/operation.go]: Done.

**Review by cheyang — APPROVED** [review]
/lgtm
/approve

**Pull request merged**